source/geometry_model/box.cc

```diff
@@ -260,6 +260,183 @@ namespace aspect
+              temp_surface_y.push_back(local_surface_height[1][i]);
+            }
+            else
+              if(local_surface_height[0][i] != local_surface_height[0][i-1])
+              {
+                 temp_surface_x.push_back(local_surface_height[0][i]);
+                 temp_surface_y.push_back(local_surface_height[1][i]);
+              }
+          }
+
+          for (unsigned int p=1; p<Utilities::MPI::n_mpi_processes(this->get_mpi_communicator()); ++p)
+          {
+            // First, find out the size of the array a process wants to send.
+            MPI_Status status;
+            MPI_Probe(p, 42, this->get_mpi_communicator(), &status);
+            int incoming_size = 0;
+            MPI_Get_count(&status, MPI_DOUBLE, &incoming_size);
+
+            // Resize the array so it fits whatever the process sends.
+            std::vector<std::vector<double>> temporary_surface(2, std::vector<double>());
+
+            for (unsigned int i=0; i<temporary_surface.size(); ++i)
+                temporary_surface[i].resize(incoming_size);
+
+            for (unsigned int i=0; i<temporary_surface.size(); ++i)
+              MPI_Recv(&temporary_surface[i][0], incoming_size, MPI_DOUBLE, p, 42, this->get_mpi_communicator(), &status);
+
+            for (unsigned int i=0; i<temporary_surface[1].size(); ++i)
+              {
+                if(i==0)
+                {
+                  if(temp_surface_x.size() > 0)
+                  {
+                    if(temporary_surface[0][i] != temp_surface_x[temp_surface_x.size() - 1])
+                    {
+                      temp_surface_x.push_back(temporary_surface[0][i]);
+                      temp_surface_y.push_back(temporary_surface[1][i]);
+                    }
+                  else
+                    {
+                      temp_surface_x.push_back(temporary_surface[0][i]);
+                      temp_surface_y.push_back(temporary_surface[1][i]);
+                    }
+                  }
+                }
+                else
+                {
+                  if(temporary_surface[0][i] != temporary_surface[0][i-1])
+                  {
+                    temp_surface_x.push_back(temporary_surface[0][i]);
+                    temp_surface_y.push_back(temporary_surface[1][i]);
+                  }
+                }
+              }
+          }
+
+            double vector_size = temp_surface_x.size();
+            MPI_Bcast(&vector_size, 1, MPI_DOUBLE, 0, this->get_mpi_communicator());
+            MPI_Bcast(&temp_surface_x[0], vector_size, MPI_DOUBLE, 0, this->get_mpi_communicator());
+            MPI_Bcast(&temp_surface_y[0], vector_size, MPI_DOUBLE, 0, this->get_mpi_communicator());
+        }
+        else
+        {
+          for (unsigned int i=0; i<local_surface_height.size(); i++)
+              MPI_Ssend(&local_surface_height[i][0], local_surface_height[1].size(), MPI_DOUBLE, 0, 42, this->get_mpi_communicator());
```

Comment:
And here you want to use the corresponding isend function: https://dealii.org/developer/doxygen/deal.II/namespaceUtilities_1_1MPI.html#a82095f4e891152bab36ebfea1423620b

Reply:
I've kept it as is for now, but I can update it to isend in the future when dealii 9.5 is required.